### PR TITLE
Feature/bms system

### PR DESCRIPTION
**Pull request by angrycompany16**
Added BMS system feature hehe

Some known limitations:
* The system doesn't detect battery packs continuously, only on startup
* The Freya diagnostics are for some reason currently not working :/ DevOps pls fix

### CODE REVIEW

File: sensors/bms/bms/freya_bms_node.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+class FreyaBMSNode(Node):
+    """
+        Publishes Freya's BMS data to ROS2.
+
+        Methods:
+        ---------
+        publish_bms_data(self) -> None
+            publises BMS data from BMS system to ROS2 node.
+        create_key_value_pair(key: str, value) -> KeyValue:
+            creates KeyValue object from supplied key and value
```
Comment (PizzaAllTheWay) [inline sensors/bms/bms/freya_bms_node.py:17]: What does the KeyValue object part of the code do in this node, does it serve a purpose? Is it somehow published to diagnostics or anywhere else?
I need context, me no understand what that part of code end goal is lol

File: sensors/bms/package.xml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>bms</name>
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
+  <maintainer email="karate.martynas1@gmail.com">vortex</maintainer>
```
Comment (PizzaAllTheWay) [inline sensors/bms/package.xml:7]: lol yes, it is I that has written all the code, I take all the credit for the beutifull code hehehehe >:)

File: sensors/bms/setup.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    install_requires=['setuptools'],
+    zip_safe=True,
+    maintainer='vortex',
+    maintainer_email='karate.martynas1@gmail.com',
+    description='TODO: Package description',
+    license='TODO: License declaration',
+    tests_require=['pytest'],
+    entry_points={
+        'console_scripts': [
+            'bms_publisher = bms.freya_bms_node:main'
```
Comment (PizzaAllTheWay) [inline sensors/bms/setup.py:26]: could you give it the same name as the node, that way people can easily identify which node they want to run withouth checking setup.py file every time they want to launch the code, just a good practice for less confusion :)
ie
freya_bms_node = bms.freya_bms_node:main

alternatively change node name to include publisher to make it even clearer, ie freya_bms_publisher_node.py, and then in setup.py write:
freya_bms_publisher_node = bms.freya_bms_publisher_node:main

File: sensors/bms/launch/bms_launch.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+from launch import LaunchDescription
+from launch_ros.actions import Node
+
+def generate_launch_description():
+    return LaunchDescription([
+        Node(
+            package='bms',
+            executable='bms_publisher'
```
Comment (PizzaAllTheWay) [inline sensors/bms/launch/bms_launch.py:8]: Read what I comented in setup.py file, this should be changed to freya_bms_node or freya_bms_publisher_node :)

**Review by PizzaAllTheWay — CHANGES_REQUESTED** [review]
Verry well written code :D
(You should teach me how to write this clean code lol XD)

There are some changes and stuff I am woundering about, read my coments bellow. All of that stuff is trivial though, so I don't think you will have an issue with it :)

Also verry important that you resolve merge conflicts, those should not be there O-O

Link the BMS wiki page plz

Verry well written README.md file

Good job <3

File: sensors/bms/bms/freya_bms_node.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
+import rclpy
+from rclpy.node import Node
+import rclpy.logging
+from sensor_msgs.msg import BatteryState
```
[the remark this answers is not in the document]
Comment (EirikKolas) [reply sensors/bms/bms/freya_bms_node.py:4]: sensor_msgs is a standard ros library

[the remark this answers is not in the document]
Comment (PizzaAllTheWay) [reply sensors/bms/bms/freya_bms_node.py:4]: ah oki ty <3

File: sensors/bms/bms/freya_bms.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
+import subprocess
+import rclpy.logging
+
+#TODO: use ros2 logging instead of python print
```
[the remark this answers is not in the document]
Comment (angrycompany16) [reply sensors/bms/bms/freya_bms.py]: fixed :)

File: sensors/bms/bms/freya_bms_node.py (hunk 1 above)
Comment (angrycompany16) [reply sensors/bms/bms/freya_bms_node.py]: The KeyValue object is needed for publishing to the /diagnostics topic. Basically the /diagnostics topic takes a message of type DiagnosticArray, which is an array of DiagnosticStatus messages. The DiagnosticStatus message contains a field  called values, which stores key-value pairs using the KeyValue object, so we can represent stuff like Voltage: 22.0 and stuff

File: sensors/bms/package.xml (hunk 2 above)
Comment (angrycompany16) [reply sensors/bms/package.xml]: teehee

File: sensors/bms/setup.py (hunk 3 above)
Comment (angrycompany16) [reply sensors/bms/setup.py]: fixed :D renamed the script as well

File: sensors/bms/launch/bms_launch.py (hunk 4 above)
Comment (angrycompany16) [reply sensors/bms/launch/bms_launch.py]: :+1:

**Pull request merged**